examples/android/clientcache/app/src/main/java/io/grpc/clientcacheexample/SafeMethodCachingInterceptor.java

```diff
@@ -0,0 +1,297 @@
+  }
+
+  private static int DEFAULT_MAX_AGE_SECONDS = 3600;
+
+  private static final Metadata.Key<String> CACHE_CONTROL_KEY =
+      Metadata.Key.of("cache-control", Metadata.ASCII_STRING_MARSHALLER);
+
+  private static final Splitter CACHE_CONTROL_SPLITTER =
+      Splitter.on(',').trimResults().omitEmptyStrings();
+
+  private final Cache internalCache;
+  private final int defaultMaxAge;
+
+  private SafeMethodCachingInterceptor(Cache cache, int defaultMaxAge) {
+    this.internalCache = cache;
+    this.defaultMaxAge = defaultMaxAge;
+  }
+
+  @Override
+  public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+      final MethodDescriptor<ReqT, RespT> method, final CallOptions callOptions, Channel next) {
+    // Currently only unary methods can be marked safe, but check anyways.
+    if (!method.isSafe() || method.getType() != MethodDescriptor.MethodType.UNARY) {
+      return next.newCall(method, callOptions);
+    }
+
+    final String fullMethodName = method.getFullMethodName();
+
+    return new ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(
+        next.newCall(method, callOptions)) {
+      private Listener<RespT> interceptedListener;
+      private Key requestKey;
+      private boolean cacheResponse = true;
+      private boolean cacheOptionsCausedCancellation;
+
+      @Override
+      public void start(Listener<RespT> responseListener, Metadata headers) {
+        interceptedListener =
+            new ForwardingClientCallListener.SimpleForwardingClientCallListener<RespT>(
+                responseListener) {
+              private Deadline deadline;
+              private int maxAge = -1;
+
+              @Override
+              public void onHeaders(Metadata headers) {
+                Iterable<String> cacheControlHeaders = headers.getAll(CACHE_CONTROL_KEY);
+                if (cacheResponse && cacheControlHeaders != null) {
+                  for (String cacheControlHeader : cacheControlHeaders) {
+                    for (String directive : CACHE_CONTROL_SPLITTER.split(cacheControlHeader)) {
+                      if (directive.equalsIgnoreCase("no-cache")) {
+                        cacheResponse = false;
+                        break;
+                      } else if (directive.equalsIgnoreCase("no-store")) {
+                        cacheResponse = false;
+                        break;
+                      } else if (directive.equalsIgnoreCase("no-transform")) {
+                        cacheResponse = false;
+                        break;
+                      } else if (directive.toLowerCase().startsWith("max-age")) {
+                        String[] parts = directive.split("=");
+                        if (parts.length == 2) {
+                          try {
+                            maxAge = Integer.parseInt(parts[1]);
+                          } catch (NumberFormatException e) {
+                            continue;
```

Comment:
this seems as least loggable.  If someone misformats the number, they should at least be able to find out about it

Reply:
Logged